### ISSUE TITLE
State variables with automatic differentiation

### ISSUE THREAD

**Issue by adtzlr** [adtzlr/felupe#882]
The current implementation of state variables together with automatic differentiation is fine as long as the required math-ops are supported by the AD-backend. To overcome these limitations, it would be good to create some kind of a pre-evaluation-function to update the state variables **outside and before** the evaluation of the strain energy function. This is computationally more costly but gives more flexibility.

$$ \psi = \psi(F, \zeta) $$

Pre-evaluation of the updated state-variable(s):

$$ \zeta= \eta(F, \zeta_n) $$

followed by the evaluation of the (gradient and hessian of the) strain energy density function.

```python
import jax
import jax.numpy as jnp
import numpy as np

# deformation gradient tensor and old state variables
F = np.eye(3) + np.arange(9).reshape(3, 3) / 10
ζ = np.arange(6)

def statevars(F, ζ):
    C = F.T @ F
    i, j = np.triu_indices(3)
    return C[i, j]

def W(F, ζ):
    J = jnp.linalg.det(F)
    C = F.T @ F
    return 1 / 2 * (J**(-2/3) * jnp.trace(C) - 3)
    
dWdF = jax.grad(W)
d2WdFdF = jax.jacfwd(dWdF)

# first update the state variables
ζ_new = statevars(F, ζ)

# then the gradient and the hessian of the strain energy density function
P = dWdF(F, ζ_new)
A = d2WdFdF(F, ζ_new)
```
